harness/determined/estimator/_estimator_trial.py

```diff
@@ -270,19 +271,44 @@ def control_loop(self) -> None:
             logging.debug(f"Received wkld {wkld.kind} with args {args}.")
 
             if wkld.kind == workload.Workload.Kind.RUN_STEP:
-                # Store values for the training loop.
-                self.num_batches = wkld.num_batches
-                self.train_response_func = response_func
-                # Break out of the control loop so that the train process
-                # re-enters the train_and_evaluate() loop.
-                break
+                try:
+                    # Store values for the training loop.
+                    self.num_batches = wkld.num_batches
+                    self.train_response_func = response_func
+                    # Break out of the control loop so that the train process
+                    # re-enters the train_and_evaluate() loop.
+                    break
```

Comment:
Um, I don't think it's possible for these three lines of code to raise a `det.InvalidHP` exception.

This is the tricky part of our plugin-wrapping-a-plugin architecture; in this callback, we interrupt the main training loop to do non-training things, then our callback releases control back to the main loop, and some time later the main loop will re-enter this callback's control_loop.

That means to capture exceptions that occur during training, you'll have to do them somewhere above the scope of the estimator training loop; not inside the estimator training loop.

I actually think what @shiyuann suggested of using the push architecture for pushing checkpoints would massively simplify this.  And handing exceptions anywhere else in a way that lets you keep the harness in-tact enough to receive another message from the master asking for a checkpoint seems so hard that I wouldn't bother... I'd just say "if you raise this exception outside of evaluation you should not expect a checkpoint", and then after shiyuan's push api lands you can clean that up.

Reply:
@rb-determined-ai What I have right now is the checkpoint and validation APIs, to make them work with the searcher and trial/experiment, there is still a bunch of work to do. For now, I think it's fine to just raise an exception.

Reply:
cool.  In that case we should probably let invalidHP exceptions during training in estimators *not* result in a checkpoint.  I think that would be too much work, and in the near future there will be an easy way to do it.